### ISSUE TITLE
`weave launch` hangs on IPAM claim

### ISSUE THREAD

**Issue by awh** [weaveworks/weave#1083]
@pacaletx encountered the following problem attempting to dynamically grow a weave 0.11.2 network ([IRC transcript](https://botbot.me/freenode/weavenetwork/2015-07-06/?msg=43773908&page=1)):

On `$HOST1`
````
weave launch -iprange 10.2.3.0/24 -initpeercount 1
weave launch-dns "10.9.1.1/24" -debug
weave launch-proxy --with-dns --with-ipam
````

Later, attempt to add `$HOST2`
````
weave launch -iprange 10.2.3.0/24 -initpeercount 2
````
with the intent of following up with `launch-dns`/`launch-proxy` and `weave connect $HOST1`. This is not possible as the `weave launch` on $HOST2 hangs:

````
Allocator range [10.2.3.0-10.2.4.0) awaiting consensus: Nodes known: 1, Quorum size: 2
Pending requests:
  Claim 1f38974edf82148f65b1a721003a89f708e922a2846a280e8f5bdb87353882e8 -> 10.2.3.1
Allocator default subnet: 10.2.3.0/24
````

@bboreham diagnosed that $HOST2 had extant application containers with `ethwe` interfaces addressed inside the IPAM range; `weave launch` was therefore waiting on IPAM quorum to reclaim these addresses, a state which would never be reached as the two routers were not yet connected. User confirmed that a `weave connect $HOST1` in a separate session on $HOST2 allowed the `weave launch` to complete.

Possible workarounds:

* Ensure no application containers are running when weave is launched
* Specify IP address to second launch
* Run `weave connect` asynchronously

all of which are inconvenient in other ways.

Finally, it was not clear why `weave launch` was stuck (similar concerns are echoed [here](https://github.com/weaveworks/weave/issues/1023#issuecomment-118781742))

**Comment by rade** [comment]
Could we make the HTTP API invoked by `ipam_claim` just queue up the claim request and return success (assuming all preliminary checks succeed)?

**Comment by bboreham** [comment]
The thing you'd lose is the ability to notify the user that another peer now owned one of the IP addresses.

Which is not hugely important compared to hanging the whole thing.

**Comment by awh** [comment]
> The thing you'd lose is the ability to notify the user that another peer now owned one of the IP addresses.

Does this scenario currently cause the launch to abort? If not, then I suppose sending a message to the router logs on failure isn't much different, particularly when the router is being launched automatically from a script.

Have we considered `detach`ing the offending local container if we join the network and find that the address is in use elsewhere? It seems drastic, but the alternative is leaving the user to debug the resulting ARP ping-pong on their own... at least detaching it would result in a hard failure.

**Comment by rade** [comment]
> Have we considered detaching the offending local container if we join the network and find that the address is in use elsewhere?

That would have to be done in the script - the router knows next to nothing about containers or interfaces - thus preventing the async outcome we are after here.

**Comment by awh** [comment]
>  the router knows next to nothing about containers or interfaces

At the moment. weaveDNS (and presumably the router when #1065 lands) listens for container events already - is this so bad?

**Comment by awh** [comment]
It doesn't have to be the router that does the actual detach - IPAM could emit an event, and we could have some kind of policy engine in a different container that takes action...

**Comment by bboreham** [comment]
At a higher level, you know that two endpoints have the same IP address, but you don't know which one is better to detach.  Absent a much more sophisticated engine, best to just make it possible for the user to know.

**Comment by rade** [comment]
Let's please not turn this issue into a complete re-architecture.

**Comment by bboreham** [comment]
Note that the start-everything `weave launch` command will still hang trying to get an address for weaveDNS. But a replacement weaveDNS that doesn't need an address is in progress.

**Comment by awh** [comment]
Merged into master.
